File: beautifulplanet/safepaw/services/router/internal/publisher/publisher.go
// =============================================================
// NOPEnclaw Router — Outbound Publisher
// =============================================================
// Publishes processed messages to the outbound Redis stream.
// The Gateway reads from this stream and delivers to WebSocket clients.
//
// Uses the same XADD pattern as the Gateway's inbound publisher,
// with MAXLEN ~10000 to prevent unbounded stream growth.
// =============================================================

package publisher

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	goredis "github.com/redis/go-redis/v9"
)

// OutboundMessage is the structure published to the outbound stream.
// Must match what the Gateway's outboundReader expects.
type OutboundMessage struct {
	MessageID string `json:"message_id"`
	SessionID string `json:"session_id"`
	Content   string `json:"content"`
	Timestamp int64  `json:"timestamp"`
}

// Publisher writes messages to the outbound Redis stream.
type Publisher struct {
	client         *goredis.Client
	outboundStream string
	maxLen         int64
	maxOutbound    int // Max serialized message size in bytes (0 = no limit)
}

// NewPublisher creates a publisher that shares the given Redis client.
func NewPublisher(client *goredis.Client, outboundStream string) *Publisher {
	return &Publisher{
		client:         client,
		outboundStream: outboundStream,
		maxLen:         10000,
	}
}

// NewPublisherWithConnection creates a Publisher with its own Redis connection.
// Use this when the publisher needs an independent connection pool.
func NewPublisherWithConnection(addr, password string, db int, outboundStream string, maxOutbound int) (*Publisher, error) {
	client := goredis.NewClient(&goredis.Options{
		Addr:         addr,
		Password:     password,
		DB:           db,
		PoolSize:     10,
		MinIdleConns: 2,
		MaxRetries:   3,
		DialTimeout:  5 * time.Second,
		ReadTimeout:  3 * time.Second,
		WriteTimeout: 3 * time.Second,
		PoolTimeout:  4 * time.Second,
	})

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := client.Ping(ctx).Err(); err != nil {
		client.Close()
		return nil, fmt.Errorf("publisher redis connection failed: %w", err)
	}

	log.Printf("[PUBLISHER] Connected to %s, stream=%s", addr, outboundStream)

	return &Publisher{
		client:         client,
		outboundStream: outboundStream,
		maxLen:         10000,
		maxOutbound:    maxOutbound,
	}, nil
}

// Publish writes a message to the outbound stream.
// Returns the Redis stream ID assigned to the message.
func (p *Publisher) Publish(ctx context.Context, msg *OutboundMessage) (string, error) {
	data, err := json.Marshal(msg)
	if err != nil {
		return "", fmt.Errorf("failed to marshal outbound message: %w", err)
	}

	// Reject oversized messages before writing to Redis.
	// Without this, a runaway agent response could write 50MB to the stream
	// and the Gateway would try to send it over a WebSocket.
	if p.maxOutbound > 0 && len(data) > p.maxOutbound {
		return "", fmt.Errorf("outbound message too large: %d bytes (max %d)", len(data), p.maxOutbound)
	}

	result, err := p.client.XAdd(ctx, &goredis.XAddArgs{
		Stream: p.outboundStream,
		MaxLen: p.maxLen,
		Approx: true,
		Values: map[string]interface{}{
			"data": string(data),
		},
	}).Result()

	if err != nil {
		return "", fmt.Errorf("XADD to %s failed: %w", p.outboundStream, err)
	}

	return result, nil
}

// Close shuts down the publisher's Redis connection.
func (p *Publisher) Close() error {
	return p.client.Close()
}

File: beautifulplanet/safepaw/services/agent/src/publisher.ts
// =============================================================
// NOPEnclaw Agent — Outbound Publisher
// =============================================================
// Publishes processed messages to the outbound Redis stream.
// The Gateway reads from this stream and delivers to WebSocket
// clients.
//
// Message format MUST match what the Gateway expects:
//   { message_id, session_id, content, timestamp }
//
// Uses XADD with MAXLEN ~10000 — same as Router's publisher.
// =============================================================

import Redis from "ioredis";
import { Config } from "./config";

/**
 * OutboundMessage — what the Gateway expects on the outbound stream.
 * Must match the Go struct in router/internal/publisher/publisher.go:
 *
 *   type OutboundMessage struct {
 *       MessageID string `json:"message_id"`
 *       SessionID string `json:"session_id"`
 *       Content   string `json:"content"`
 *       Timestamp int64  `json:"timestamp"`
 *   }
 */
export interface OutboundMessage {
  message_id: string;
  session_id: string;
  content: string;
  timestamp: number;
}

/**
 * Publisher writes messages to the outbound Redis stream.
 * Uses its own Redis connection (separate from the consumer).
 */
export class Publisher {
  private client: Redis;
  private outboundStream: string;
  private maxLen: number;
  private maxOutboundSize: number;

  constructor(cfg: Config) {
    const [host, portStr] = cfg.redisAddr.split(":");
    const port = parseInt(portStr || "6379", 10);

    this.outboundStream = cfg.outboundStream;
    this.maxLen = 10000;
    this.maxOutboundSize = cfg.maxOutboundSize;

    this.client = new Redis({
      host,
      port,
      password: cfg.redisPassword,
      db: cfg.redisDB,
      maxRetriesPerRequest: 3,
      lazyConnect: true,
    });
  }

  /**
   * Connect to Redis and verify connectivity.
   */
  async connect(): Promise<void> {
    await this.client.connect();
    console.log(
      `[PUBLISHER] Connected, stream=${this.outboundStream}`
    );
  }

  /**
   * Publish writes a message to the outbound stream.
   * Returns the Redis stream ID assigned to the entry.
   *
   * Format: XADD <stream> MAXLEN ~ 10000 * data <json>
   * The Gateway reads entries and expects {"data": "<json>"}.
   */
  async publish(msg: OutboundMessage): Promise<string> {
    const data = JSON.stringify(msg);

    console.log(
      `[PUBLISHER] Publishing outbound: msg_id=${msg.message_id} ` +
        `session=${msg.session_id} size=${data.length} bytes`
    );

    // Reject oversized messages before writing to Redis.
    // Without this, a runaway response could fill the stream with
    // multi-MB entries that the Gateway tries to send over WebSocket.
    if (this.maxOutboundSize > 0 && data.length > this.maxOutboundSize) {
      const err = `Outbound message too large: ${data.length} bytes (max ${this.maxOutboundSize})`;
      console.error(`[PUBLISHER] REJECTED: ${err}`);
      throw new Error(err);
    }

    const streamId = await this.client.xadd(
      this.outboundStream,
      "MAXLEN",
      "~",
      String(this.maxLen),
      "*",
      "data",
      data
    );

    if (!streamId) {
      const err = `XADD to ${this.outboundStream} returned null`;
      console.error(`[PUBLISHER] FAILED: ${err}`);
      throw new Error(err);
    }

    console.log(
      `[PUBLISHER] ✔ Published: msg_id=${msg.message_id} stream_id=${streamId}`
    );

    return streamId;
  }

  /**
   * Health check — verifies Redis connectivity.
   */
  async healthCheck(): Promise<void> {
    const result = await this.client.ping();
    if (result !== "PONG") {
      throw new Error(`Redis health check failed: ${result}`);
    }
  }

  /**
   * Close the Redis connection.
   */
  async close(): Promise<void> {
    await this.client.quit();
    console.log("[PUBLISHER] Connection closed");
  }
}
